### PR TITLE
ipn/ipnlocal, net/tsdial: plumb routes into tsdial and use them in UserDial

### LINKED ISSUE #4529: cmd/tailscale: 'nc' subcommand doesn't work on macOS
`tailscale nc derp10 22` should show an SSH banner greeting upon connect.

It just hangs on macOS:

```
goroutine 1 [select]:
net/http.(*persistConn).roundTrip(0x1400015aea0, 0x1400020e280)
	/Users/bradfitz/sdk/go1.18/src/net/http/transport.go:2620 +0x810
net/http.(*Transport).roundTrip(0x140000cc000, 0x140000ca100)
	/Users/bradfitz/sdk/go1.18/src/net/http/transport.go:594 +0x6bc
net/http.(*Transport).RoundTrip(0x14000240160?, 0x104bb0ca8?)
	/Users/bradfitz/sdk/go1.18/src/net/http/roundtrip.go:17 +0x20
net/http.send(0x140000ca100, {0x104bb0ca8, 0x140000cc000}, {0x1048cc630?, 0x8?, 0x0?})
	/Users/bradfitz/sdk/go1.18/src/net/http/client.go:252 +0x4fc
net/http.(*Client).send(0x14000094660, 0x140000ca100, {0x22?, 0x0?, 0x0?})
	/Users/bradfitz/sdk/go1.18/src/net/http/client.go:176 +0x9c
net/http.(*Client).do(0x14000094660, 0x140000ca100)
	/Users/bradfitz/sdk/go1.18/src/net/http/client.go:725 +0x6ec
net/http.(*Client).Do(...)
	/Users/bradfitz/sdk/go1.18/src/net/http/client.go:593
tailscale.com/client/tailscale.DoLocalRequest(0x104b90a20?)
	/Users/bradfitz/src/tailscale.com/client/tailscale/tailscale.go:102 +0x78
tailscale.com/client/tailscale.DialTCP({0x104bb3b68, 0x1400013c008}, {0x16b777ada, 0x6}, 0x97?)
	/Users/bradfitz/src/tailscale.com/client/tailscale/tailscale.go:477 +0x3a4
tailscale.com/cmd/tailscale/cli.runNC({0x104bb3b68, 0x1400013c008}, {0x14000094580, 0x2, 0x0?})
	/Users/bradfitz/src/tailscale.com/cmd/tailscale/cli/nc.go:50 +0x13c
github.com/peterbourgon/ff/v3/ffcli.(*Command).Run(0x0?, {0x104bb3b68?, 0x1400013c008?})
	/Users/bradfitz/go/pkg/mod/github.com/peterbourgon/ff/v3@v3.1.2/ffcli/command.go:153 +0x144
github.com/peterbourgon/ff/v3/ffcli.(*Command).Run(0x140000c4000?, {0x104bb3b68?, 0x1400013c008?})
	/Users/bradfitz/go/pkg/mod/github.com/peterbourgon/ff/v3@v3.1.2/ffcli/command.go:157 +0xf4
tailscale.com/cmd/tailscale/cli.Run({0x14000094570?, 0x3?, 0x140000a3f28?})
	/Users/bradfitz/src/tailscale.com/cmd/tailscale/cli/cli.go:203 +0x7e4
main.main()
	/Users/bradfitz/src/tailscale.com/cmd/tailscale/tailscale.go:24 +0x184
```

The `UserDial` call in `localapi.go` is likely hanging:

```go
func (h *Handler) serveDial(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		http.Error(w, "POST required", http.StatusMethodNotAllowed)
		return
	}
	const upgradeProto = "ts-dial"
	if !strings.Contains(r.Header.Get("Connection"), "upgrade") ||
		r.Header.Get("Upgrade") != upgradeProto {
		http.Error(w, "bad ts-dial upgrade", http.StatusBadRequest)
		return
	}
	hostStr, portStr := r.Header.Get("Dial-Host"), r.Header.Get("Dial-Port")
	if hostStr == "" || portStr == "" {
		http.Error(w, "missing Dial-Host or Dial-Port header", http.StatusBadRequest)
		return
	}
	hijacker, ok := w.(http.Hijacker)
	if !ok {
		http.Error(w, "make request over HTTP/1", http.StatusBadRequest)
		return
	}

	addr := net.JoinHostPort(hostStr, portStr)
	outConn, err := h.b.Dialer().UserDial(r.Context(), "tcp", addr)
	if err != nil {
		http.Error(w, "dial failure: "+err.Error(), http.StatusBadGateway)
		return
	}
	defer outConn.Close()
```

/cc @maisem

### PR DESCRIPTION
**Pull request by nickkhyl**
We'd like to use tsdial.Dialer.UserDial instead of SystemDial for DNS over TCP. This is primarily necessary to properly dial internal DNS servers accessible over Tailscale and subnet routes. However, to avoid issues when switching between Wi-Fi and cellular, we need to ensure that we don't retain connections to any external addresses on the old interface. Therefore, we need to determine which dialer to use internally based on the configured routes.

This plumbs routes and localRoutes from router.Config to tsdial.Dialer, and updates UserDial to use either the peer dialer or the system dialer, depending on the network address and the configured routes.

Updates tailscale/corp#18725
Fixes #4529

### CODE REVIEW

File: net/tsdial/tsdial.go [hunk 1]
```diff
@@ -76,6 +75,7 @@ type Dialer struct {
 	dnsCache         *dnscache.MessageCache // nil until first non-empty SetExitDNSDoH
 	nextSysConnID    int
 	activeSysConns   map[int]net.Conn // active connections not yet closed
+	routes           art.Table[bool]  // true indicates routes that point into the Tailscale interface
```
Comment (maisem) [inline net/tsdial/tsdial.go:78]: Prefer to use bart.Table these days as it's more memory efficient

Comment (nickkhyl) [reply net/tsdial/tsdial.go]: Updated to use `bart.Table` instead of `art.Table`. Thanks!

**Comment by nickkhyl** [comment]
Updated commit message:
Fixes #4529

File: net/tsdial/tsdial.go [hunk 2]
```diff
@@ -129,6 +129,20 @@ func (d *Dialer) SetExitDNSDoH(doh string) {
 	}
 }
 
+// SetRoutes configures the dialer to dial the specified routes via Tailscale,
+// and the specified localRoutes using the default interface.
```
Comment (bradfitz) [inline net/tsdial/tsdial.go:133]: are the localRoutes even needed? It's just localhost addresses and WSL2's weirdo local address. Neither will be coming from control.

File: net/tsdial/tsdial.go [hunk 3]
```diff
@@ -129,6 +129,20 @@ func (d *Dialer) SetExitDNSDoH(doh string) {
...
+func (d *Dialer) SetRoutes(routes, localRoutes []netip.Prefix) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.routes = bart.Table[bool]{}
```
Comment (bradfitz) [inline net/tsdial/tsdial.go:137]: you could make the table first, and then acquire the mutex once you're ready to assign to d.routes.

File: net/tsdial/tsdial.go (hunk 2 above)
Comment (nickkhyl) [reply net/tsdial/tsdial.go:133]: I think they're needed, as they also include LAN addresses if `ipn.Prefs.ExitNodeAllowLANAccess` is enabled:
https://github.com/tailscale/tailscale/blob/ce8969d82b5df9eff4f3ad71f6781c73e4481bf0/ipn/ipnlocal/local.go#L4196

File: net/tsdial/tsdial.go (hunk 3 above)
Comment (nickkhyl) [reply net/tsdial/tsdial.go:137]: I've been considering that. But `bart.Table` contains a `sync.Once`, and I believe it should not be copied after its first used by any operation on the table.

Comment (bradfitz) [reply net/tsdial/tsdial.go:137]: you should be storing it as a pointer, not a value.

Comment (bradfitz) [reply net/tsdial/tsdial.go:137]: (note that all its methods are on the pointer receiver... and see https://go.dev/wiki/CodeReviewComments#receiver-type)

File: net/tsdial/tsdial.go (hunk 2 above)
Comment (bradfitz) [reply net/tsdial/tsdial.go:133]: ah, okay

File: net/tsdial/tsdial.go (hunk 3 above)
Comment (nickkhyl) [reply net/tsdial/tsdial.go]: Done.

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -2259,6 +2259,10 @@ const (
 
 	// NodeAttrSuggestExitNodeUI allows the currently suggested exit node to appear in the client GUI.
 	NodeAttrSuggestExitNodeUI NodeCapability = "suggest-exit-node-ui"
+
+	// NodeAttrUserDialUseRoutes makes UserDial use either the peer dialer or the system dialer,
+	// depending on the destination address and the configured routes.
+	NodeAttrUserDialUseRoutes NodeCapability = "user-dial-routes"
```
Comment (maisem) [inline tailcfg/tailcfg.go:2265]: bump capver too

Comment (nickkhyl) [reply tailcfg/tailcfg.go:2266]: Done, thanks!

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -3636,6 +3636,12 @@ func (b *LocalBackend) authReconfig() {
 	}
 	b.logf("[v1] authReconfig: ra=%v dns=%v 0x%02x: %v", prefs.RouteAll(), prefs.CorpDNS(), flags, err)
 
+	if b.netMap.SelfNode.HasCap(tailcfg.NodeAttrUserDialUseRoutes) {
```
Comment (maisem) [inline ipn/ipnlocal/local.go:3639]: i think you want to use `b.netmap.HasCap`

File: net/tsdial/tsdial.go [hunk 6]
```diff
@@ -75,7 +76,8 @@ type Dialer struct {
 	exitDNSDoHBase   string                 // non-empty if DoH-proxying exit node in use; base URL+path (without '?')
 	dnsCache         *dnscache.MessageCache // nil until first non-empty SetExitDNSDoH
 	nextSysConnID    int
-	activeSysConns   map[int]net.Conn // active connections not yet closed
+	activeSysConns   map[int]net.Conn  // active connections not yet closed
+	routes           *bart.Table[bool] // or nil if UserDial should not use routes. `true` indicates routes that point into the Tailscale interface
```
Comment (maisem) [inline net/tsdial/tsdial.go:80]: grabbing and releasing a mutex here might be too expensive for dns queries. I think you should store this as an `atomic.Pointer`

File: ipn/ipnlocal/local.go (hunk 5 above)
Comment (bradfitz) [reply ipn/ipnlocal/local.go:3639]: Also `b.netMap` is guarded by `b.mu` which isn't held right now, so this is a race.

But `nm` holds `b.netMap`, so use `nm.HasCap`

File: net/tsdial/tsdial.go (hunk 6 above)
Comment (nickkhyl) [reply net/tsdial/tsdial.go]: Good point, thank you! Updated to use an `atomic.Pointer`.

File: ipn/ipnlocal/local.go (hunk 5 above)
Comment (nickkhyl) [reply ipn/ipnlocal/local.go]: Done, thanks!

File: net/tsdial/tsdial.go [hunk 7]
```diff
@@ -75,7 +77,8 @@ type Dialer struct {
 	exitDNSDoHBase   string                 // non-empty if DoH-proxying exit node in use; base URL+path (without '?')
 	dnsCache         *dnscache.MessageCache // nil until first non-empty SetExitDNSDoH
 	nextSysConnID    int
-	activeSysConns   map[int]net.Conn // active connections not yet closed
+	activeSysConns   map[int]net.Conn                 // active connections not yet closed
+	routes           atomic.Pointer[bart.Table[bool]] // or nil if UserDial should not use routes. `true` indicates routes that point into the Tailscale interface
```
Comment (maisem) [inline net/tsdial/tsdial.go:81]: nit: as this is no longer protected by `mu` move it before the block so that its clearer that its not protected.

**Review by maisem — APPROVED** [review]

**Review by bradfitz — APPROVED** [review]

**Pull request merged**